File: backend/app/schemas/trace.py
"""
与主计划 TraceStep 契约及前端 `lib/types/trace.ts` 对齐的最小 OpenAPI 模型。
持久化层仍为 JSON；此处用于响应校验与文档显式化。
"""

from typing import Any

from pydantic import BaseModel, ConfigDict, Field


class TraceStepMeta(BaseModel):
    model_config = ConfigDict(extra="allow")

    model: str | None = None
    step_type: str | None = None
    tokens: int | None = None
    cost_estimate: float | None = None
    latency: float | None = None
    retryCount: int | None = None
    prompt: str | None = None
    label: str | None = None
    tool: dict[str, Any] | None = None
    rag: dict[str, Any] | None = None
    memory: list[str] | None = None


class TraceStep(BaseModel):
    model_config = ConfigDict(extra="allow")

    id: str
    type: str = Field(
        ...,
        description="三大类：thought / action / observation；扩展类型仍用 string",
    )
    content: str
    seq: int | None = Field(
        default=None,
        description="与 trace delta 游标对齐；缺省时由持久化层按序填充",
    )
    meta: TraceStepMeta | None = None


def parse_trace_steps(steps: list[dict]) -> list[TraceStep]:
    """将 SQLite 中 JSON 反序列化后的 dict 转为显式模型（供 OpenAPI 与校验）。"""
    return [TraceStep.model_validate(s) for s in steps]

File: frontend/lib/types/trace.ts
/**
 * 与 `.cursor/plans` 中 TraceStep 契约及后端 SSE `trace` 事件对齐。
 * `type` 三大类便于 Flow；细分子类走 `meta.step_type`。
 */
export type TraceStepKind =
  | "thought"
  | "action"
  | "observation"
  /** 显式类型或与 meta.tool / meta.rag 联用 */
  | "tool"
  | "rag";

export type TraceStepMeta = {
  model?: string;
  step_type?: string;
  prompt_tokens?: number | null;
  completion_tokens?: number | null;
  usage_source?: "provider" | "estimated" | "legacy" | string;
  planning_provider_attempted?: boolean;
  planning_provider_used?: boolean;
  tool_registry_profile?: string;
  tool_registry_provider_source?: string;
  allowed_tool_names?: string[];
  allowed_tool_labels?: string[];
  tool?: {
    name: string;
    label?: string;
    kind?: string;
    semantic_kind?: string | null;
    semantic_family?: string | null;
    input?: unknown;
    output?: unknown;
    output_preview?: unknown;
    status: "running" | "done" | "error";
    supports_result_preview?: boolean;
    effective_result_preview_keys?: string[];
    effective_result_output_keys?: string[];
    retry_count?: number;
    error?: string | null;
  };
  rag?: { chunks: string[]; knowledge_base_id?: string };
  memory?: string[];
  latency?: number;
  tokens?: number | null;
  cost_estimate?: number | null;
  retryCount?: number;
  prompt?: string;
  /** 扩展：自定义展示标题 */
  label?: string;
};

export type TraceStepPayload = {
  id: string;
  type: TraceStepKind | string;
  content: string;
  /** 与 delta 游标对齐的序号，缺省时由索引推导 */
  seq?: number;
  meta?: TraceStepMeta;
};
